lib/Conversion/ImportVerilog/Expressions.cpp

```diff
@@ -361,12 +361,56 @@ struct ExprVisitor {
     for (auto *operand : expr.operands()) {
       auto value = context.convertExpression(*operand);
       if (!value)
-        return {};
+        continue;
+      value = convertToSimpleBitVector(value);
       operands.push_back(value);
     }
     return builder.create<moore::ConcatOp>(loc, operands);
   }
 
+  // Handle replications.
+  Value visit(const slang::ast::ReplicationExpression &expr) {
+    auto type = context.convertType(*expr.type);
+    if (isa<moore::VoidType>(type))
+      return {};
+
+    auto value = context.convertExpression(expr.concat());
+    if (!value)
+      return {};
+    return builder.create<moore::ReplicateOp>(loc, type, value);
+  }
+
+  // Handle single bit selections.
+  Value visit(const slang::ast::ElementSelectExpression &expr) {
+    auto type = context.convertType(*expr.type);
+    auto value = context.convertExpression(expr.value());
+    auto lowBit = context.convertExpression(expr.selector());
+
+    if (!value || !lowBit)
+      return {};
+    return builder.create<moore::ExtractOp>(loc, type, value, lowBit);
+  }
```

Comment:
What happens if the expression is an array, not an integer? I think it would make sense to have different operations for the different types that the bit select expressions can operate on. For example, accessing a single bit in an integer is likely going to be quite a bit different than accessing a single element in an array.

It would be good if this code looked at the type of the value being bit-selected into, handled the specific types that it understands, and errors out on all other cases. (You want the IR to have very clear semantics and reject any input that you can't really handle early on.)

How about adding a `moore.array_get` operation, similar to `hw.array_get`, which can operate on packed/unpacked arrays? And leave `moore.extract` to operate on integers. Then this code would check the type of `expr.value()` and generate `extract` for integers, `array_get` for arrays, and an error otherwise.

Reply:
Please let me know if I misunderstood. 
We can support selection from `int a;`(like `a[1]` or `a[var]`), `arr[3:1];`(like `arr[1]` or `arr[var]`), and the type of `arr` can be `predefinedWidthType`(like int, byte, and so on), `moore.packedArray` and `moore.unpackedArray`. But we need to handle the special situation when a variable as `lowBit` like `logic [3:0] l1;    logic [1:0] l2;   logic l3 = l2[l1];`

Reply:
What happens if you do something like `a[b:c]`, where `b` and `c` are both variables?

Reply:
`a[b:c]` will trigger an error: reference to non-constant variable 'c' is not allowed in a constant expression. This error will be generated by either variable `b` or `c`.